tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,711 @@
+			if _, err := f.WriteAt(w.buf[:8], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		w.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func sequenceFilesMap(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+func closeAllFromMap(cs map[int]io.Closer) error {
+	var merr tsdb_errors.MultiError
+
+	for _, c := range cs {
+		merr.Add(c.Close())
+	}
+	return merr.Err()
+}
+
+// CorruptionErr is an error that's returned when corruption is encountered.
+type CorruptionErr struct {
+	Dir     string
+	Segment int
+	Err     error
+}
+
+func (e *CorruptionErr) Error() string {
+	return fmt.Sprintf("corruption in segment %s: %s", segmentFile(e.Dir, e.Segment), e.Err.Error())
+}
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the segment file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (w *HeadReadWriter) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(w.bs))
+	for seg := range w.bs {
+		seqs = append(seqs, seg)
+	}
+	sort.Ints(seqs)
+	for _, seq := range seqs {
+		bs := w.bs[seq]
+		sliceLen := bs.Len()
+		idx := HeadSegmentHeaderSize
+		for idx < sliceLen {
+			if sliceLen-idx < 30 {
+				return &CorruptionErr{
+					Dir:     w.dirFile.Name(),
+					Segment: seq,
+					Err:     errors.Errorf("segment doesn't include enough bytes to read the chunk header - required:%v, available:%v", idx+35, sliceLen),
+				}
+			}
+
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
+				return &CorruptionErr{
+					Dir:     w.dirFile.Name(),
+					Segment: seq,
+					Err:     err,
+				}
+			}
+
+			idx++ // Skip encoding.
+			// Skip the data.
+			dataLen, n := binary.Uvarint(bs.Range(idx, idx+MaxChunkLengthFieldSize))
+			idx += n + int(dataLen)
+		}
+		if idx > sliceLen {
+			// It should be equal to the slice length. Else it means the last chunks had less bytes.
+			return &CorruptionErr{
+				Dir:     w.dirFile.Name(),
+				Segment: seq,
+				Err:     errors.Errorf("segment doesn't include enough bytes to read the last chunk data - required:%v, available:%v", idx, sliceLen),
+			}
+		}
+	}
+
+	return nil
+}
+
+// Repair deletes all the segments after the one which had the corruption
+// (including the corrupt file).
+func (w *HeadReadWriter) Repair(originalErr error) error {
```

Comment:
Hm it looks like it deletes stuff not repairs (:

Reply:
That is how current WAL Repair works too, it deletes everything after the corruption point.